src/tp_engine.cpp

```diff
@@ -60,6 +265,76 @@ void TPEngine::UpdateDAGAfterAddNodePairByParsimony(const NNIOperation &nni_op)
   Failwith("Currently no implementation.");
 }
 
+EdgeId TPEngine::BestEdgeAdjacentToNode(const NodeId node_id,
+                                        const Direction direction) const {
+  const auto node = dag_.GetDAGNode(node_id);
+  EdgeId best_edge_id = EdgeId(NoId);
+  size_t best_tree_source = GetTreeCount();
+  for (const auto focal_clade : {SubsplitClade::Left, SubsplitClade::Right}) {
+    for (const auto rootward_node_id : node.GetNeighbors(direction, focal_clade)) {
+      const auto edge_id = dag_.GetEdgeIdx(rootward_node_id, node_id);
+      if (best_tree_source > tree_source_[edge_id.value_]) {
+        best_tree_source = tree_source_[edge_id.value_];
+        best_edge_id = edge_id;
+      }
+    }
+  }
+  return best_edge_id;
+}
+
+// ** Partial Vector Operations
+
+void TPEngine::Set(const PVId dest_id, const PVId src_id) {
+  auto &pvs = likelihood_pvs_;
+  pvs.GetPV(dest_id).array() = pvs.GetPV(src_id).array();
+}
+
+void TPEngine::Multiply(const PVId dest_id, const PVId src1_id, const PVId src2_id) {
+  auto &pvs = likelihood_pvs_;
+  pvs.GetPV(dest_id).array() = pvs.GetPV(src1_id).array() * pvs.GetPV(src2_id).array();
+  // rescaling_counts_(op.dest_) =
```

Comment:
Do we need to raise a little issue about rescaling?

Reply:
Yes, good point.  I added one the issue.

Reply:
I don't see a difference in the code-- perhaps we want to make a note there? 

In any case it seems like long-term this code should get refactored out to reduce code duplication when we are beyond proof of concept stage.

Reply:
Removed code and added note.  Added the note about refactoring it with GPEngine's code to the issue.